app/models/work_url.rb

```diff
@@ -0,0 +1,73 @@
+class WorkUrl < ApplicationRecord
+  belongs_to :work
+
+  VARIANTS = %w[original minimal no_www with_www encoded decoded minimal_no_protocol_no_www].freeze
+
+  def self.build_url_attributes(url)
+    url = UrlFormatter.new(url)
+
+    VARIANTS.to_h { |s| [s.to_sym, url.send(s)] }
+  end
+
+  def self.prepare(work, url)
+    work_url = WorkUrl.find_or_initialize_by(work_id: work.id)
+
+    work_url.assign_attributes(self.build_url_attributes(url))
+
+    work.work_url = work_url
+
+    work_url
+  end
```

Comment:
I'm sure there's a more idiosyncratic Rails syntax for that but I'm drawing a blank at the moment.